restapi/api_test.go

```diff
@@ -1 +1,429 @@
+	}
+}
+
+// TestReadListOfClustersEmptyResponse check the method ReadListOfClusters
+func TestReadListOfClustersEmptyResponse(t *testing.T) {
+	// start a local HTTP server
+	server := mockedHttpServer(func(responseWriter http.ResponseWriter, request *http.Request) {
+		checkURL(t, request, "/api/v1/client/cluster")
+	})
+	// close the server when test finishes
+	defer server.Close()
+
+	api := restapi.NewRestAPI(server.URL)
+
+	// perform REST API call against mocked HTTP server
+	_, err := api.ReadListOfClusters()
+	if err == nil {
+		t.Fatal("Error is expected to be returned")
+	}
+}
+
+// TestReadListOfClustersWrongJSON check the method ReadListOfClusters
+func TestReadListOfClustersWrongJSON(t *testing.T) {
+	// start a local HTTP server
+	server := mockedHttpServer(func(responseWriter http.ResponseWriter, request *http.Request) {
+		checkURL(t, request, "/api/v1/client/cluster")
+		// send response to be tested later
+		writeBody(responseWriter, `this is not proper JSON`)
+	})
+	// close the server when test finishes
+	defer server.Close()
+
+	api := restapi.NewRestAPI(server.URL)
+
+	// perform REST API call against mocked HTTP server
+	_, err := api.ReadListOfClusters()
+	if err == nil {
+		t.Fatal("Error is expected to be returned")
+	}
+}
+
+// TestReadListOfClustersResponseError check the method ReadListOfClusters
+func TestReadListOfClustersResponseError(t *testing.T) {
+	// start a local HTTP server
+	server := httptest.NewServer(http.NotFoundHandler())
+	// close the server when test finishes
+	defer server.Close()
+
+	api := restapi.NewRestAPI(server.URL)
+
+	// perform REST API call against mocked HTTP server
+	_, err := api.ReadListOfClusters()
+	if err == nil {
+		t.Fatal("Error is expected to be returned")
+	}
+}
+
+// TestReadListOfTriggersEmptyList check the method ReadListOfTriggers
+func TestReadListOfTriggersEmptyList(t *testing.T) {
+	// start a local HTTP server
+	server := mockedHttpServer(func(responseWriter http.ResponseWriter, request *http.Request) {
+		checkURL(t, request, "/api/v1/client/trigger")
+		// send response to be tested later
+		writeBody(responseWriter, `{"status":"ok"}`)
+	})
+	// close the server when test finishes
+	defer server.Close()
+
+	api := restapi.NewRestAPI(server.URL)
+
+	// perform REST API call against mocked HTTP server
+	triggers, err := api.ReadListOfTriggers()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(triggers) != 0 {
+		t.Fatal("Expected empty list of triggers")
+	}
+}
+
+// TestReadListOfTriggersOneTrigger check the method ReadListOfTriggers
+func TestReadListOfTriggersOneTrigger(t *testing.T) {
+	// start a local HTTP server
+	server := mockedHttpServer(func(responseWriter http.ResponseWriter, request *http.Request) {
+		checkURL(t, request, "/api/v1/client/trigger")
+		responseAsString := `
+		{
+			"triggers": [{"ID":0,"Name":"Name","Type":"must-gather"}],
+			"status":"ok"
+		}`
+		// send response to be tested later
+		writeBody(responseWriter, responseAsString)
```

Comment:
Out of curiosity - how would this behave if `writeBody` wrote a string formatted like that (with all the whitespace) instead of a the byte slice? 
Would the `ResponseWriter` handle it? Or does Go handle that?

Reply:
no it should be byte slice, so the conversion via `byte()` needs to be there

Reply:
OK, I looked at the docs and I realize now that the `ResponseWriter.Write` func only takes a `[]byte`, so I'll ask a different question - how would the body look if we used `io.WriteString(responseWriter)`?
Or is usage of `io.WriteString` discouraged for these purposes?

Reply:
both can be used IIRC